### PR TITLE
DRILL-4783: flatten operator should not throw exception if there is empty resultset

### PR DESCRIPTION
**Pull request by chunhui-shi**
returned for underlying operator which is convert_from or other complex functions.

### CODE REVIEW

**Comment by chunhui-shi** [comment]
@jaltekruse Hi Jason, could you help to review this change? Thanks.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java [hunk 1]
```diff
@@ -129,8 +129,14 @@ private void setFlattenVector() {
...
       final TypedFieldId typedFieldId = incoming.getValueVectorId(popConfig.getColumn());
       final MaterializedField field = incoming.getSchema().getColumn(typedFieldId.getFieldIds()[0]);
-      final RepeatedValueVector vector = RepeatedValueVector.class.cast(incoming.getValueAccessorById(
-          field.getValueClass(), typedFieldId.getFieldIds()).getValueVector());
+      RepeatedValueVector vector = null;
+      try {
+        vector = RepeatedValueVector.class.cast(incoming.getValueAccessorById(
+            field.getValueClass(), typedFieldId.getFieldIds()).getValueVector());
+      } catch(ClassCastException ex) {
+        //if cast fail
```
Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java]: If the input to flatten happens to be a scalar value, i.e (int 100), what's the desired result? The previous behavior is to hit ClassCastException, which might need improve the error message.  With your change, flatten(100) will return a single value of (100), or just an empty repeated map?

Comment (chunhui-shi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java]: Thanks Jinfeng for the review. The fix will return empty result if it is flatten(100).

Since this fix is to address the issue when there is no incoming record, flatten above convert_from should not throw exception. Meanwhile it is better to keep other behavior unchanged unless we have strong reason to change it. Based on this understanding, I am updating the change to still throw the same cast exception if flatten is not getting a RepeatedValueVector to work on.

**Comment by jaltekruse** [comment]
The original behavior was not by design, it was a bug. I would just detect the case where a scalar is being flattened and throw an appropriate user facing message like "Flatten does not support inputs of non-list values".

The tests you have included here don't check their results. I would also suggest adding a test for the new negative case, checking for the appropriate message.

If you want to more precisely control the input to your test, I would recommend taking a look at the operator unit test system I checked in and demonstrated at the hangout last month. This will allow you to write more targeted tests for this case. Feel free to ask questions if the usage is not clear from the examples in this class.

https://github.com/apache/drill/blob/master/exec/java-exec/src/test/java/org/apache/drill/exec/physical/unit/BasicPhysicalOpUnitTest.java

**Comment by chunhui-shi** [comment]
Thanks @jaltekruse for the review. I updated exception text and added test case following your suggestion.

File: exec/java-exec/src/test/java/org/apache/drill/exec/nested/TestFastComplexSchema.java [hunk 2]
```diff
@@ -80,4 +80,33 @@ public void test4() throws Exception {
             "                       AND r.r_regionkey = 4)) t \n" +
             "ORDER  BY t.f.name");
   }
+
+  @Test //DRILL-4783 when resultset is empty, don't throw exception.
+  public void test5() throws Exception {
+    test("alter session set `planner.enable_hashjoin` = false");
```
Comment (jinfengni) [inline exec/java-exec/src/test/java/org/apache/drill/exec/nested/TestFastComplexSchema.java]: Any reason why you need set these two options?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java [hunk 3]
```diff
@@ -129,11 +129,21 @@ private void setFlattenVector() {
     try {
       final TypedFieldId typedFieldId = incoming.getValueVectorId(popConfig.getColumn());
       final MaterializedField field = incoming.getSchema().getColumn(typedFieldId.getFieldIds()[0]);
-      final RepeatedValueVector vector = RepeatedValueVector.class.cast(incoming.getValueAccessorById(
-          field.getValueClass(), typedFieldId.getFieldIds()).getValueVector());
+      RepeatedValueVector vector = null;
+      try {
+        vector = RepeatedValueVector.class.cast(incoming.getValueAccessorById(
+            field.getValueClass(), typedFieldId.getFieldIds()).getValueVector());
+      } catch(ClassCastException ex) {
```
Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java]: In stead of catch ClassCastException, you may check if the input value vector is an instance of RepeatedValueVector.  Using try{} catch{} seems a bit over-kill for this case.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java [hunk 4]
```diff
@@ -129,11 +129,21 @@ private void setFlattenVector() {
...
+          throw UserException.unsupportedError(ex).message("Flatten does not support inputs of non-list values.").build(logger);
+        }
+        //when incoming recordCount is 0, don't throw exception since the type being seen here is not solid
+        logger.error("setFlattenVector cast failed and recordcount is 0, create empty vector anyway.");
+        vector = new RepeatedMapVector(field, oContext.getAllocator(), null);
+      }
       flattener.setFlattenField(vector);
     } catch (Exception ex) {
-      throw UserException.unsupportedError(ex).message("Trying to flatten a non-repeated field.").build(logger);
+      throw UserException.unsupportedError(ex).message("Flatten does not support inputs of non-list values.").build(logger);
```
Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java]: UserException may be thrown twice? Line 138 & Line 146.

Comment (chunhui-shi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java]: The two catch blocks are parallel. The UserException in line 138 wont be caught and thrown again in line 146.

**Comment by chunhui-shi** [comment]
Update based on Jinfeng's comment. Jinfeng, could you check if I addressed your concern? Thanks.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java [hunk 5]
```diff
@@ -267,10 +276,21 @@ private TransferPair getFlattenFieldTransferPair(FieldReference reference) {
...
-      // vvIn may be null because of fast schema return for repeated list vectors
-      if (vvIn != null) {
-        tp = vvIn.getTransferPair(reference.getAsNamePart().getName(), oContext.getAllocator());
+      try {
+        final ValueVector vvIn = RepeatedValueVector.class.cast(flattenField).getDataVector();
+        // vvIn may be null because of fast schema return for repeated list vectors
+        if (vvIn != null) {
+          tp = vvIn.getTransferPair(reference.getAsNamePart().getName(), oContext.getAllocator());
+        }
+      } catch(ClassCastException ex) {
```
Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java]: We can apply similar changes above here ( catch ClassCastException -> instancOf check), also, logger same message.

**Comment by jinfengni** [comment]
+1
Overall LGTM.

@jaltekruse , do you have any further comment? thx.

**Comment by jaltekruse** [comment]
One small last comment. The method testNoResult() doesn't actually verify that they query has no results, it just ignores whatever the query returns. I would recommend instead using the expectsEmptyResultSet() method on the test builder and marking this method deprecated, with a comment about the method on the builder as the preferred interface for such tests.

With that small fix I'm +1

**Comment by jinfengni** [comment]
Thanks, @jaltekruse . I'll make the change you recommended, and squashed into Chunhui's patch.

**Comment by jinfengni** [comment]
LGTM.

**Comment by jinfengni** [comment]
I picked the latest commit, and am running the regression. Will merge to master if everything runs fine.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java [hunk 6]
```diff
@@ -267,6 +276,15 @@ private TransferPair getFlattenFieldTransferPair(FieldReference reference) {
...
+      if ( !(flattenField instanceof RepeatedValueVector) ) {
+        if(incoming.getRecordCount() != 0) {
+          throw UserException.unsupportedError().message("Flatten does not support inputs of non-list values.").build(logger);
+        }
+        logger.error("Cannot cast {} to RepeatedValueVector", flattenField);
+        //when incoming recordCount is 0, don't throw exception since the type being seen here is not solid
+        ValueVector vv = new RepeatedMapVector(flattenField.getField(), oContext.getAllocator(), null);
+        tp = RepeatedValueVector.class.cast(vv).getTransferPair(reference.getAsNamePart().getName(), oContext.getAllocator());
+      }
       final ValueVector vvIn = RepeatedValueVector.class.cast(flattenField).getDataVector();
```
Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java]: Looks like an else is missing here. Otherwise, the first new unit test (with empty input for flatten) will fail with ClassCastException.
